Add expanded view on code snippets

esonhey opened the pull request:
Refactored the Code component 
 - Removed `element` property since there was no usage of it in our codebase
 - Removed `isColorSyntax` property since we can handle it with the `language` property, if the language property is passed so we colorize it
 - Removed the padding property, it's better to add the spacing where we are using it instead of having it there.
 - Added two missing properties in story book stories so we can test it there better.

Added an ExpandViewer component to handle the logic of expanded view generally
 - For the expand button didn't use the rebass Button component since we gonna use another styling tool soon, also I didn't use the usual .css format since I didn't see anywhere that we have used the CSS file.

Added Expanded View capability for Code component

esonhey committed:
Add missing props in story

esonhey committed:
Add arrows maximize icon svg

esonhey committed:
Add BreifViewer compoennt

esonhey committed:
Add expandabality to Code Snippet

esonhey committed:
Rename BriefViewer to ExpandViewer

esonhey committed:
Remove old padding property from story

giardiv commented:
Task linked: [CU-1ptc2at Add expanded view on code snippets](https://app.clickup.com/t/1ptc2at)

esonhey commented on src/components/code/index.tsx:
```diff
@@ -14,41 +15,249 @@ import styles, {
 }
 
+const defaultPopupProps = {
+  width: '700px',
```
@giardiv 
what would be the default width of the popup window if the code line is too long?
I have set it to 700px by default, forgot to ask about it before raising the PR

esonhey replied:
changed to have it as relative to window width

esonhey committed:
Remove redundant default props

esonhey committed:
Refactor to reuse the almost same logic of codes

esonhey committed:
Change the default width and height of popup

esonhey committed:
Fix downloading actual content on breif
It were downloading the brief content on breif overview now fixed
to download/copy the actuall content but show the brief

snqb commented on src/index.ts:
```diff
@@ -175,6 +176,7 @@ export {
   Divider,
   PerformanceCard,
   Slider,
+  ExpandViewer,
```
let's export the props too

esonhey replied:
done

snqb commented on src/components/expand-viewer/index.tsx:
```diff
@@ -0,0 +1,100 @@
+import { expandViewButtonStyle } from './expand-viewer.styles';
+import { ITheme } from '../../theme/types';
+
+// ExpandViewer
```
nit: do we need this comment? It doesn't seem to serve any value, as the file name and the component name speak for themselves

esonhey replied:
yeah forgot to remove it, it's here since first I developed this component inside the Code component then later move it out to its own module.

snqb commented on src/components/expand-viewer/index.tsx:
```diff
@@ -0,0 +1,100 @@
+  BriefComponent,
+  popupProps,
+}) => {
+  const [expanded, setExpanded] = useState(false);
```
there is `usePopup`, which is very suitable for cases like this.
Have you considered using it?

esonhey replied:
yeah checked that but it's not suitable for this case since the second returned is the handleToggle which we don't need here.

snqb commented on src/components/expand-viewer/index.tsx:
```diff
@@ -0,0 +1,100 @@
+            <Box flexGrow={1}>
+              <Subtitle>{title}</Subtitle>
+            </Box>
+            <IconButton
+              icon={IconName.cross}
+              intent="ghost"
+              onClick={() => setExpanded(false)}
+            ></IconButton>
```
can we use `<IconButton />` here?
No children here

snqb commented on src/components/expand-viewer/expand-viewer.styles.ts:
```diff
@@ -0,0 +1,17 @@
+  borderRadius: '50px',
+  borderWidth: '1px',
+  borderStyle: 'solid',
+  /**
+   * since it's not possible to use :hover selector in inline React style I had to handle it manually,
+   * there could be added a css style which avoided since there is no css files in the project
+   */
```
is it? We have  plenty of usages of `:hover` in the project

e.g. `src/components/row/group/row-group.styles.ts`

esonhey replied:
yeah I see, since @giardiv told me to not go with rebass stuff I tried making my own button component, and unfortunately the :hover selector doesn't not work in plain react inline CSS . but I think we should still keep with rebass until we pick a new stack

snqb replied:
yeah, just use rebass stuff until we get rid off it. We just don't know when that happens yet, as it's a very big change to be made
until that, rebass is doing its job

snqb commented on src/components/expand-viewer/expand-viewer.styles.ts:
```diff
@@ -0,0 +1,17 @@
+   * there could be added a css style which avoided since there is no css files in the project
+   */
+  backgroundColor: isHovered ? theme.colors.white : theme.colors.primaryShade2,
+  borderColor: theme.colors.primary,
```
you don't need to access `theme` programmatically
since we use `emotion-theming`, you can just but `borderColor: 'primary'`. This is the case for every color usage

esonhey replied:
same reason to avoid using styling from emotion CSS. gonna update to use rebass instead.

snqb commented on src/components/code/index.tsx:
```diff
@@ -62,77 +224,19 @@ const Code: FC<CodeProps> = ({
+      >
+        {icons.copy}
+        <Value ml="5px" mt="1px">
+          {!copied ? 'copy' : 'copied'}
```
nit: avoid `!` where you can just flip the conditions
it makes code 1% more difficult to read, which we don't need

esonhey replied:
good one! will fix it.
actually, I was just refactoring, and moving out the component didn't touch the inside logic that much.

snqb commented on src/components/code/index.tsx:
```diff
@@ -62,77 +224,19 @@ const Code: FC<CodeProps> = ({
+    <Box>
+      <Button
+        intent="ghost"
+        sx={{ ...buttonsStyles }}
```
why spread here? Can it be `sx={buttonStyles}`?

esonhey replied:
good point, old stuff but gonna fix this ;)

snqb commented on src/components/code/index.tsx:
```diff
@@ -14,41 +15,202 @@ import styles, {
+    return (
+      <ExpandViewer
+        title="Code snippet expanded view"
+        /** currently merging with built-in javascript spread, if we need deep merging we may use lodash merge function instead */
```
I think it's alright this way
props should never be passed in non-shallow way

snqb commented on src/components/code/index.tsx:
```diff
@@ -14,41 +15,202 @@ import styles, {
+const DownloadButton: FC<
+  Pick<CodeProps, 'content' | 'downloadCallback' | 'title'>
+> = ({ content, downloadCallback, title }) => {
+  const download = () => {
+    if (downloadCallback) {
+      downloadCallback();
```
I know that this is a piece of old code, but since you're in the context: don't we need to call the `downloadCallback` in all cases? Now it's either save to file or call callback. 
Shouldn't it just save even if the callback is there?
@giardiv

esonhey replied:
I see in the code somewhere we are using just the `donwloadButton` prop without specifying `downloadCallback`.
https://github.com/logicalclocks/hopsworks-front/blob/1025acf20ec578558d8ca509694651d6bb5bc158/src/pages/project/training-dataset/forms/Features/FeatureTransformations/ViewTransformationPopup.tsx#L34-L45

snqb commented on src/components/code/index.tsx:
```diff
@@ -14,41 +15,202 @@ import styles, {
+}) => {
+  const [copied, setCopied] = useState(false);
+
   const useCodeKey = async () => {
```
I know this is old code too, but it breaks conventions not to use `use-` prefix for anything besides hooks
can you rename this in the PR?

esonhey replied:
good one, sure will update the name

snqb commented on src/components/code/index.tsx:
```diff
@@ -14,41 +15,202 @@ import styles, {
+  );
+};
+
+export default Code;
+
+// Normal Code Component
+interface CodeSnippetProps
```
is it a pattern you like with specifying default export and then below writing other components that are part of the main one?
it looks neat, but I'm not 100% sure

esonhey replied:
Yes, it's the kinda a practice I like. if the components and the main default exported component is not big I would keep them in the same file. 
It depends on how you guys are comfortable with it, if not I can separate them to their own file?

snqb reviewed (changes requested):
actually, a very good job
I like the approach with the `ExpandViewer` 
but there are quite a few problems, mostly these are caused by you being new to the codebase 😺 

but I left the comments

esonhey committed:
Renaming copy button handler

esonhey committed:
Minor refactoring

esonhey committed:
Refac minor stuff in expand viewer

esonhey committed:
Change contentToShow to optional property
Change logic to show content if contendToShow is not provided

esonhey committed:
Export ExpandViewerProps from the root index

esonhey committed:
Refactor expand button to use rebass button

giardiv commented:
It's pixel-perfect, nicely done

Waiting for last code approval bf merging it

snqb reviewed (approved):
very good!
